src/objective/hinge.cu

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright 2018 by Contributors
+ * \file hinge.cc
+ * \brief Provides an implementation of the hinge loss function
+ * \author Henry Gouk
+ */
+#include <xgboost/objective.h>
+#include "../common/math.h"
+#include "../common/transform.h"
+#include "../common/common.h"
+#include "../common/span.h"
+#include "../common/host_device_vector.h"
+
+namespace xgboost {
+namespace obj {
+
+#if defined(XGBOOST_USE_CUDA)
+DMLC_REGISTRY_FILE_TAG(hinge_obj_gpu);
+#endif
+
+struct HingeObjParam : public dmlc::Parameter<HingeObjParam> {
+  int n_gpus;
+  int gpu_id;
+  DMLC_DECLARE_PARAMETER(HingeObjParam) {
+    DMLC_DECLARE_FIELD(n_gpus).set_default(0).set_lower_bound(0)
+        .describe("Number of GPUs to use for multi-gpu algorithms.");
+    DMLC_DECLARE_FIELD(gpu_id)
+        .set_lower_bound(0)
+        .set_default(0)
+        .describe("gpu to use for objective function evaluation");
+  }
+};
+
+class HingeObj : public ObjFunction {
+ public:
+  HingeObj() = default;
+
+  void Configure(
+      const std::vector<std::pair<std::string, std::string> > &args) override {
+    param_.InitAllowUnknown(args);
+    devices_ = GPUSet::All(param_.n_gpus).Normalised(param_.gpu_id);
+    label_correct_.Resize(devices_.IsEmpty() ? 1 : devices_.Size());
+  }
+
+  void GetGradient(const HostDeviceVector<bst_float> &preds,
+                   const MetaInfo &info,
+                   int iter,
+                   HostDeviceVector<GradientPair> *out_gpair) override {
+    CHECK_NE(info.labels_.Size(), 0U) << "label set cannot be empty";
+    CHECK_EQ(preds.Size(), info.labels_.Size())
+        << "labels are not correctly provided"
+        << "preds.size=" << preds.Size()
+        << ", label.size=" << info.labels_.Size();
+
+    const bool is_null_weight = info.weights_.Size() == 0;
+    const size_t ndata = preds.Size();
+    out_gpair->Resize(ndata);
```

Comment:
If a `HostDeviceVector` with an empty distribution is resized, memory will be allocated for it on the host. If it is later resharded, this host memory allocation will remain.

Therefore, it is better to reshard a `HostDeviceVector` before resizing it, rather than the other way around.

Reply:
Thanks for the details.